File: WPF/Controls/Squaring.vb
Namespace Controls

    Public Class Squaring
        Inherits ContentControl

        Shared Sub New()
            DefaultStyleKeyProperty.OverrideMetadata(GetType(Squaring), New FrameworkPropertyMetadata(GetType(Squaring)))
        End Sub

        Protected Overrides Function MeasureOverride(ByVal Constraint As Size) As Size
            Dim Dimension = Math.Min(Constraint.Width, Constraint.Height)
            Dim Size = New Size(Dimension, Dimension)

            Dim Content = Me.Content
            Content?.Measure(Size)

            If Me.FillAvailableSpace Then
                If Double.IsPositiveInfinity(Dimension) Then
                    Return New Size()
                End If
                Return Size
            End If

            Dimension = 0.0
            If Content IsNot Nothing Then
                Size = Content.DesiredSize
                Dimension = Math.Max(Size.Width, Size.Height)
            End If

            Return New Size(Dimension, Dimension)
        End Function

        Protected Overrides Function ArrangeOverride(arrangeBounds As Size) As Size
            Dim Content = Me.Content
            If Content Is Nothing Then
                Return arrangeBounds
            End If

            Dim Rect As Rect
            If arrangeBounds.Height < arrangeBounds.Width Then
                Rect = New Rect((arrangeBounds.Width - arrangeBounds.Height) / 2, 0, arrangeBounds.Height, arrangeBounds.Height)
            Else
                Rect = New Rect(0, (arrangeBounds.Height - arrangeBounds.Width) / 2, arrangeBounds.Width, arrangeBounds.Width)
            End If

            Content.Arrange(Rect)

            Return arrangeBounds
        End Function

#Region "FillAvailableSpace Property"
        Public Shared ReadOnly FillAvailableSpaceProperty As DependencyProperty = DependencyProperty.Register("FillAvailableSpace", GetType(Boolean), GetType(Squaring), New PropertyMetadata(True, AddressOf FillAvailableSpace_Changed, AddressOf FillAvailableSpace_Coerce))

        Private Shared Function FillAvailableSpace_Coerce(ByVal D As DependencyObject, ByVal BaseValue As Object) As Object
            'Dim Self = DirectCast(D, Squaring)

            'Dim Value = DirectCast(BaseValue, Boolean)

            Return BaseValue
        End Function

        Private Shared Sub FillAvailableSpace_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, Squaring)

            'Dim OldValue = DirectCast(E.OldValue, Boolean)
            'Dim NewValue = DirectCast(E.NewValue, Boolean)

            Self.InvalidateMeasure()
        End Sub

        Public Property FillAvailableSpace As Boolean
            Get
                Return DirectCast(Me.GetValue(FillAvailableSpaceProperty), Boolean)
            End Get
            Set(ByVal value As Boolean)
                Me.SetValue(FillAvailableSpaceProperty, value)
            End Set
        End Property
#End Region

    End Class

End Namespace

File: WPF/Controls/Squaring.cs
using System.Windows;
using System;

namespace Ks
{
    namespace Common.Controls
    {
        public class Squaring : ContentElement
        {
            static Squaring()
            {
                DefaultStyleKeyProperty.OverrideMetadata(typeof(Squaring), new FrameworkPropertyMetadata(typeof(Squaring)));
            }

            protected override Size MeasureOverride(Size Constraint)
            {
                var Dimension = Math.Min(Constraint.Width, Constraint.Height);
                var Size = new Size(Dimension, Dimension);

                var Content = this.Content;
                Content?.Measure(Size);

                if (this.FillAvailableSpace)
                {
                    if (double.IsPositiveInfinity(Dimension))
                        return new Size();
                    return Size;
                }

                Dimension = 0.0;
                if (Content != null)
                {
                    Size = Content.DesiredSize;
                    Dimension = Math.Max(Size.Width, Size.Height);
                }

                return new Size(Dimension, Dimension);
            }

            protected override Size ArrangeOverride(Size arrangeBounds)
            {
                var Content = this.Content;
                if (Content == null)
                    return arrangeBounds;

                Rect Rect;
                if (arrangeBounds.Height < arrangeBounds.Width)
                    Rect = new Rect((arrangeBounds.Width - arrangeBounds.Height) / 2, 0, arrangeBounds.Height, arrangeBounds.Height);
                else
                    Rect = new Rect(0, (arrangeBounds.Height - arrangeBounds.Width) / 2, arrangeBounds.Width, arrangeBounds.Width);

                Content.Arrange(Rect);

                return arrangeBounds;
            }

            public static readonly DependencyProperty FillAvailableSpaceProperty = DependencyProperty.Register("FillAvailableSpace", typeof(bool), typeof(Squaring), new PropertyMetadata(true, FillAvailableSpace_Changed, FillAvailableSpace_Coerce));

            private static object FillAvailableSpace_Coerce(DependencyObject D, object BaseValue)
            {
                // Dim Self = DirectCast(D, Squaring)

                // Dim Value = DirectCast(BaseValue, Boolean)

                return BaseValue;
            }

            private static void FillAvailableSpace_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
            {
                var Self = (Squaring)D;

                // Dim OldValue = DirectCast(E.OldValue, Boolean)
                // Dim NewValue = DirectCast(E.NewValue, Boolean)

                Self.InvalidateMeasure();
            }

            public bool FillAvailableSpace
            {
                get
                {
                    return (bool)this.GetValue(FillAvailableSpaceProperty);
                }
                set
                {
                    this.SetValue(FillAvailableSpaceProperty, value);
                }
            }
        }
    }
}
